subsys/bluetooth/services/ans.c

```diff
@@ -0,0 +1,514 @@
+static int transmit_unread_alert(enum bt_ans_category category)
+{
+	int rc;
+
+	uint8_t cat_in = (uint8_t)category;
+
+	/* Nothing to do if notify is disabled */
+	if (!unread_alert_notify) {
+		return 0;
+	}
+
+	/* Special case: send all categories */
+	if (cat_in == BT_ANS_SEND_ALL_CATEGORY) {
+		for (int i = 0; i < BT_ANS_CAT_NUM; i++) {
+			if ((unread_sup_cat_bit_mask & BIT(i)) &&
+			    (unread_cat_enabled_map & BIT(i))) {
+				rc = notify_unread_alert_category(i);
+				if (rc < 0) {
+					return rc;
+				}
+			}
+		}
+		return 0;
+	}
+
+	/* Otherwise send just the requested category (if enabled) */
+	if (unread_cat_enabled_map & BIT(cat_in)) {
+		return notify_unread_alert_category(cat_in);
+	}
+
+	return 0;
+}
+
+int bt_ans_notify_new_alert(enum bt_ans_category category, uint8_t num_new, const char *text,
+			    int text_len)
+{
+	int ret;
+
+	uint8_t cat_in = (uint8_t)category;
+
+	/* Check if the category is supported */
+	if ((alert_sup_cat_bit_mask & (1U << cat_in)) == 0) {
+		LOG_WRN("Category %d unsupported", cat_in);
+		return -EINVAL;
+	}
+
+	/* Update the saved value */
+	ret = k_mutex_lock(&new_alert_mutex, K_FOREVER);
+	if (ret != 0) {
+		LOG_ERR("Unable to lock mutex (err: %d)", ret);
+		return ret;
+	}
+	new_alerts[cat_in].num_new_alerts = num_new;
+
+	int size = MIN(BT_ANS_MAX_TEXT_STR_SIZE - 1, text_len);
+
+	strncpy(new_alerts[cat_in].text_string, text, size);
+	new_alerts[cat_in].text_string[size] = '\0';
+	ret = k_mutex_unlock(&new_alert_mutex);
+	if (ret != 0) {
+		LOG_ERR("Unable to unlock mutex (err: %d)", ret);
+		return ret;
+	}
+
+	return transmit_new_alert(category);
+}
+
+int bt_ans_set_unread_count(enum bt_ans_category category, uint8_t unread)
+{
+	int ret;
+
+	uint8_t cat_in = (uint8_t)category;
+
+	/* Check if the category is supported */
+	if ((unread_sup_cat_bit_mask & (1U << cat_in)) == 0) {
+		LOG_WRN("Category %d unsupported", cat_in);
+		return -EINVAL;
+	}
+
+	/* Update the saved value */
+	ret = k_mutex_lock(&unread_mutex, K_FOREVER);
+	if (ret != 0) {
+		LOG_ERR("Unable to lock mutex (err: %d)", ret);
+		return ret;
+	}
+	unread_alerts[cat_in].unread_count = unread;
+	ret = k_mutex_unlock(&unread_mutex);
+	if (ret != 0) {
+		LOG_ERR("Unable to unlock mutex (err: %d)", ret);
+		return ret;
+	}
+
+	return transmit_unread_alert(category);
+}
+
+/* Callback used by bt_conn_foreach() to detect an established connection. */
+static void ans_conn_check_cb(struct bt_conn *conn, void *data)
+{
+	int err;
+	bool *has_conn = data;
+	struct bt_conn_info info;
+
+	err = bt_conn_get_info(conn, &info);
+	if (err == 0 && info.state == BT_CONN_STATE_CONNECTED) {
+		*has_conn = true;
+	}
+}
+
+int bt_ans_set_new_alert_support_category(enum bt_ans_category category, bool enable)
+{
+	uint8_t cat_in = (uint8_t)category;
+	bool has_conn = false;
+
+	bt_conn_foreach(BT_CONN_TYPE_ALL, ans_conn_check_cb, &has_conn);
+
+	if (has_conn) {
+		/* Cannot change support while connection exists */
+		return -EBUSY;
+	}
```

Comment:
Consider moving this to a `static bool` function so that it can be reused by `bt_ans_set_unread_support_category`